### PR TITLE
Remove dim and resolve the resulting problems

### PR DESCRIPTION
**Pull request by willtebbutt**
Removes `dim` as a field of `DiagEuclideanMetric` and `DenseEuclideanMetric` in favour of just using `getproperty` methods. This introduces some ambiguities, so we have to go with the following interface for construction:
```julia
UnitEuclideanMetric{Float64}(D)
DiagEuclideanMetric(::AbstractVector)
DenseEuclideanMetric(::AbstractMatrix)
```
and we've dropped the construct-with-element-from-space interface, which was really only a convenience thing anyway (as I understand it).

### CODE REVIEW

**Comment by willtebbutt** [comment]
Latest commits:
- remove element types from the `AbstractMetric`s, as they don't appear to have been necessary, and caused interface problems further down the line.
- fix an issue in `rand_momentum(::AbstractRNG, ::DiagEuclideanMetric)` whereby the `_temp` variable was being returned, and causes a bug. The fact that it works at the minute was purely a consequence of the `_temp` field being initialised to contain zeros.

As a consequence of the above, we have the following interface for metric construction:
```julia
UnitEuclideanMetric(D)
DiagEuclideanMetric(::AbstractVector)
DenseEuclideanMetric(::AbstractMatrix)
```
The only difference relate to the above is that we no longer need the type parameter to construct `UnitEuclideanMetric`.

@xukai92 this is really for review now.

**Comment by willtebbutt** [comment]
Current break seems to only happen on 1.0 and may well be related to [this](https://github.com/JuliaLang/julia/pull/31512). Beyond that, I have _literally_ no idea what might be happening.

**Comment by xukai92** [comment]
I see. Then I will ignore the check and review first.

File: README.md [hunk 1]
```diff
@@ -25,7 +25,7 @@ n_adapts = 2_000
 θ_init = randn(D)
 
 # Define metric space, Hamiltonian and sampling method
-metric = DenseEuclideanMetric(θ_init)
+metric = UnitEuclideanMetric(D)
```
Comment (xukai92) [inline README.md]: Why do we change the example of using dense?

**Comment by willtebbutt** [comment]
@xukai92 just resolved some small issues.

File: src/metric.jl [hunk 2]
```diff
@@ -32,54 +28,39 @@ struct DiagEuclideanMetric{T<:Real,A<:AbstractVector{T}} <: AbstractMetric{T}
...
-function (dem::DenseEuclideanMetric)(M⁻¹::A) where {T<:Real,A<:AbstractMatrix{T}}
-    return DenseEuclideanMetric(dem.dim, M⁻¹)
+function DenseEuclideanMetric(M⁻¹::AbstractMatrix{T}) where {T<:Real}
+    _temp = Vector{T}(undef, size(M⁻¹, 1))
+    return DenseEuclideanMetric(M⁻¹, cholesky(Symmetric(M⁻¹)).U, _temp)
 end
 
-function DenseEuclideanMetric(θ::A) where {T<:Real,A<:AbstractVector{T}}
-    dim = length(θ)
-    return DenseEuclideanMetric(dim, Matrix{T}(I, dim, dim))
```
Comment (xukai92) [inline src/metric.jl]: I suggest we provide the interface `AbstractMetric(d::Int)` to allow users initialise with identity `M⁻¹` for both `DenseEuclideanMetric` and `DiagEuclideanMetric`

File: README.md (hunk 1 above)
Comment (xukai92) [reply README.md]: If this is because of we don't have easy interface, see my comments below.

**Comment by xukai92** [comment]
Have you check type stability after the changes?

**Comment by xukai92** [comment]
By which I mean run `@code_warntype` for `sample()` function.

File: src/metric.jl (hunk 2 above)
Comment (willtebbutt) [reply src/metric.jl]: Good shout, will implement.

**Comment by willtebbutt** [comment]
> Have you check type stability after the changes?

Yes, appears to be fine.

File: README.md (hunk 1 above)
Comment (willtebbutt) [reply README.md]: It was, will revert with proposed interface below.

**Comment by xukai92** [comment]
> Yes, appears to be fine.

Awesome!

File: src/metric.jl (hunk 2 above)
Comment (xukai92) [reply src/metric.jl]: Happy to merge it when this is done.

**Comment by willtebbutt** [comment]
Right, I've created the default methods. We get `Float64`s by default, but the user can of course provide a matrix explicitly if this isn't the desired behaviour.

**Comment by xukai92** [comment]
Great work! Thanks!

**Comment by xukai92** [comment]
I'll merge it when CI passes.

**Pull request merged**